src/btree/bt_sync.c

```diff
@@ -248,6 +266,43 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
                     btree->rec_max_txn = mod->rec_max_txn;
                 if (mod != NULL && btree->rec_max_timestamp < mod->rec_max_timestamp)
                     btree->rec_max_timestamp = mod->rec_max_timestamp;
+
+                /*
+                 * Check whether the tree is lookaside btree and verify whether the page contents
+                 * are obsolete or not by checking the visibility of the page stop time pair.
+                 */
+                if (is_las) {
+                    __wt_ref_is_obsolete(session, walk, &obsolete);
+                    if (obsolete) {
+                        /*
+                         * The duplicate tree walk code expects the ref state to be in memory. We
+                         * need to get the duplicate tree walk pointer before marking the current
+                         * tree page as deleted.
+                         */
+                        WT_ERR(__sync_dup_walk(session, walk, flags, &prev));
+
+                        /*
+                         * Mark the page as deleted and also set the parent page as dirty. This is
+                         * to ensure when the parent page is checkpointing, the empty child page
+                         * will be cleaned.
+                         */
+                        previous_state = walk->state;
+                        if (!WT_PAGE_IS_INTERNAL(walk->page)) {
+                            if (!WT_REF_CAS_STATE(session, walk, previous_state, WT_REF_LOCKED))
+                                continue;
+
+                            /* Mark the page as deleted */
+                            WT_REF_SET_STATE(walk, WT_REF_DELETED);
+                            WT_ERR(__wt_page_parent_modify_set(session, walk, true));
+                        }
+
+                        /*
+                         * The duplicate tree walk pointer, prev, was obtained before we marked the
+                         * page as deleted. Continue the rest of the tree walk using that pointer.
+                         */
+                        goto skipwalk;
+                    }
+                }
```

Comment:
It looks to me like we can jump to `skipwalk` if we fail to set the `WT_REF` to `WT_REF_LOCKED`, which is a little simpler.

We also don't use `previous_state` for anything I don't see any reason to have it

Something like:
```
/*  
 * Check whether the tree is lookaside btree and verify whether the page contents
 * are obsolete or not by checking the visibility of the page stop time pair.
 */  
if (is_las) {
    __wt_ref_is_obsolete(session, walk, &obsolete);
    if (obsolete) {
        /*  
         * The duplicate tree walk code expects the ref state to be in memory. We
         * need to get the duplicate tree walk pointer before marking the current
         * tree page as deleted.
         */  
        WT_ERR(__sync_dup_walk(session, walk, flags, &prev));
        
        /*  
         * Mark the page as deleted and also set the parent page as dirty. This is
         * to ensure when the parent page is checkpointing, the empty child page
         * will be cleaned.
         */  
        if (WT_PAGE_IS_INTERNAL(walk->page) &&
            WT_REF_CAS_STATE(session, walk, walk->state, WT_REF_LOCKED)) {
            WT_REF_SET_STATE(walk, WT_REF_DELETED);
            WT_ERR(__wt_page_parent_modify_set(session, walk, true));
        }   

        /*  
         * The duplicate tree walk pointer, prev, was obtained before we marked the 
         * page as deleted. Continue the rest of the tree walk using that pointer.
         */  
        goto skipwalk;
    }   
}   
continue;
```

Reply:
I further optimized the code into a function as this function is further used when traversing internal page cleanup stage.